flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -367,50 +366,47 @@ public Task(
 
 		final String taskNameWithSubtaskAndId = taskNameWithSubtask + " (" + executionId + ')';
 
-		// Produced intermediate result partitions
-		this.producedPartitions = new ResultPartition[resultPartitionDeploymentDescriptors.size()];
+		// add metrics for buffers
+		final MetricGroup buffersGroup = metrics.getIOMetricGroup().addGroup("buffers");
 
-		int counter = 0;
+		final Optional<MetricGroup> outputGroup, inputGroup;
 
-		for (ResultPartitionDeploymentDescriptor desc: resultPartitionDeploymentDescriptors) {
-			ResultPartitionID partitionId = new ResultPartitionID(desc.getPartitionId(), executionId);
-
-			this.producedPartitions[counter] = new ResultPartition(
-				taskNameWithSubtaskAndId,
-				this,
-				jobId,
-				partitionId,
-				desc.getPartitionType(),
-				desc.getNumberOfSubpartitions(),
-				desc.getMaxParallelism(),
-				networkEnvironment.getResultPartitionManager(),
-				resultPartitionConsumableNotifier,
-				ioManager,
-				desc.sendScheduleOrUpdateConsumersMessage());
-
-			++counter;
+		// register detailed network metrics, if configured
+		if (taskManagerConfig.getConfiguration().getBoolean(TaskManagerOptions.NETWORK_DETAILED_METRICS)) {
```

Comment:
I think it does not hurt always to add only the metric group. I would suggest to move `NETWORK_DETAILED_METRICS` to `NetworkEnvironmentConfiguration`. Then we do not need `optionals` and `NetworkEnvironment.registerIn(Out)putMetrics` can decide whether to add metrics to the group based on `NetworkEnvironmentConfiguration.isDetailedMetrics()`.

Reply:
There are two reasons for using `Optional` here:

1. If we pass `Input/OutputGroup` directly, `NETWORK_DETAILED_METRICS` would be checked twice in both `Task` and `NetworkEnvironmentConfiguration`.

2. The `Input/OutputGroup` might be nullable. IMO the optional parameters in interface method might seem better than `Nullable`, also comparing `InputGroup.isPresent()` with `if (InputGroup != null)'.

Reply:
I meant not to check `NETWORK_DETAILED_METRICS` in Task and always create groups. They will just stay empty if `NetworkEnvironment` adds nothing and it should not hurt. What do you think?

Reply:
I misunderstood your previous meaning. 
Hmm it would be better to always create this group in task, then it could keep the same format parameters as other group, also not bring any hurt for users. Totally agree